os_stub/memlib/copy_mem.c

```diff
@@ -9,35 +9,68 @@
  **/
 
 #include "base.h"
+#include "library/debuglib.h"
 
 /**
- * Copies a source buffer to a destination buffer, and returns the destination buffer.
+ * Copies bytes from a source buffer to a destination buffer.
  *
- * This function copies length bytes from source_buffer to destination_buffer, and returns
- * destination_buffer.  The implementation must be reentrant, and it must handle the case
- * where source_buffer overlaps destination_buffer.
+ * This function copies "len" bytes from "src_buf" to "dst_buf".
  *
- * If length is greater than (MAX_ADDRESS - destination_buffer + 1), then ASSERT().
- * If length is greater than (MAX_ADDRESS - source_buffer + 1), then ASSERT().
+ * Asserts and returns a non-zero value if any of the following are true:
+ *   1) ("src_buf" or "dst_buf" are NULL) and "len" is greater 0.
+ *   2) "src_buf" and "dst_buf" overlap.
+ *   3) "len" is greater than "dst_len".
+ *   4) "len" or "dst_len" is greater than (MAX_ADDRESS - "dst_buf" + 1).
+ *   5) "len" or "dst_len" is greater than (MAX_ADDRESS - "src_buf" + 1).
+ *   6) "len" or "dst_len" is greater than (SIZE_MAX >> 1).
  *
- * @param  destination_buffer   A pointer to the destination buffer of the memory copy.
- * @param  source_buffer        A pointer to the source buffer of the memory copy.
- * @param  length              The number of bytes to copy from source_buffer to destination_buffer.
+ * In case of error, the "dst_buf" is left unmodifed. This behavior is different
+ * than memcopy_s, which zeros the "dst_buf" if "dst_buf" is valid.
  *
- * @return destination_buffer.
+ * @param    dst_buf   Destination buffer to copy to.
+ * @param    dst_len   Maximum length in bytes of the destination buffer.
+ * @param    src_buf   Source buffer to copy from.
+ * @param    len       The number of bytes to copy.
  *
- **/
-void *copy_mem(OUT void *destination_buffer, IN const void *source_buffer,
-               IN uintn length)
+ * @return   0 on success. non-zero on error.
+ *
+**/
+int copy_mem_s(OUT void* dst_buf, IN uintn dst_len, IN const void* src_buf, IN uintn len)
 {
-    volatile uint8_t *pointer_dst;
-    const volatile uint8_t *pointer_src;
+    volatile uint8_t* dst;
```

Comment:
This is me being a C language lawyer but when incrementing/decrementing a pointer by a certain number of bytes the type should be `char`. It just so happens that `uint8_t` is a `unsigned char`.

Reply:
Can you elaborate more on this one? I tend to favor unsigned types unless there is a need to be signed.

Reply:
I cannot understand the comment. @steven-bellock , could you please share the source of restriction? C standard? compiler?

Reply:
Signedness doesn't play a role here, it's the definition of a byte and its association with `char`. Although musl uses `unsigned char` for some reason. https://git.musl-libc.org/cgit/musl/tree/src/string/memcpy.c#n7

Reply:
I guess because of `(uintptr_t)s` it needs to be of a specific signedness.

Reply:
What is conclusion here? I am lost.

Reply:
Heh forget I brought it up. In the grand scheme of things `uint8_t` is fine since it will always be `unsigned char`.